### PR TITLE
{Monitor} refactor the logic

### PR DESCRIPTION
**Pull request by MyronFanQiu**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S169

File: src/application-insights/azext_applicationinsights/custom.py [hunk 1]
```diff
@@ -90,10 +108,19 @@ def get_component(client, application, resource_group_name):
...
         if resource_group_name:
-            return client.get(resource_group_name, application)
+            latest_client = applicationinsights_mgmt_plane_client(cmd.cli_ctx,
+                                                                  api_version='2020-02-02-preview').components
+            try:
+                return latest_client.get(resource_group_name, application)
+            except CloudError:
+                message = " Please use `az feature register " \
+                          "--name AIWorkspacePreview --namespace microsoft.insights` to register the feature"
+                logger.warning(message)
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/custom.py:122]: why do we log warning here, while raise exception in create/update?

File: src/application-insights/azext_applicationinsights/custom.py [hunk 2]
```diff
@@ -58,26 +57,45 @@ def create_or_update_component(cmd, client, application, resource_group_name, lo
...
-    if kind:
-        existing_component.kind = kind
+    from ._client_factory import applicationinsights_mgmt_plane_client
+    existing_component = None
     if workspace_resource_id is not None:
+        latest_client = applicationinsights_mgmt_plane_client(cmd.cli_ctx, api_version='2020-02-02-preview').components
+        try:
+            existing_component = latest_client.get(resource_group_name, application)
+        except CloudError as ex:
+            message = " Please use `az feature register " \
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/custom.py:80]: Define this message a constant for reuse as there are multiple places using it.

File: src/application-insights/azext_applicationinsights/custom.py [hunk 3]
```diff
@@ -58,26 +57,45 @@ def create_or_update_component(cmd, client, application, resource_group_name, lo
...
+                      "--name AIWorkspacePreview --namespace microsoft.insights` to register the feature"
+            ex.error._message = ex.error._message + message
+            raise ex
         existing_component.workspace_resource_id = workspace_resource_id or None
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/custom.py:84]: >workspace_resource_id [](start = 51, length = 21)

In which case, None will be assigned to existing_component.workspace_resource_id?

File: src/application-insights/azext_applicationinsights/custom.py [hunk 4]
```diff
@@ -58,26 +57,45 @@ def create_or_update_component(cmd, client, application, resource_group_name, lo
...
+    else:
+        existing_component = client.get(resource_group_name, application)
+    if kind:
+        existing_component.kind = kind
     if public_network_access_for_ingestion is not None:
         existing_component.public_network_access_for_ingestion = public_network_access_for_ingestion
     if public_network_access_for_query is not None:
         existing_component.public_network_access_for_query = public_network_access_for_query
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/custom.py:92]: >public_network_access_for_query [](start = 27, length = 31)

public_network_access_for_query exists in both 2018 and 2020 versions?

File: src/application-insights/azext_applicationinsights/custom.py [hunk 5]
```diff
@@ -90,10 +108,19 @@ def get_component(client, application, resource_group_name):
     return client.get(resource_group_name, application)
 
 
-def show_components(client, application=None, resource_group_name=None):
+def show_components(cmd, client, application=None, resource_group_name=None):
+    from ._client_factory import applicationinsights_mgmt_plane_client
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/custom.py:112]: What's the behavior if both application and resource_group_name are provided?

File: src/application-insights/azext_applicationinsights/custom.py (hunk 4 above)
Comment (MyronFanQiu) [reply src/application-insights/azext_applicationinsights/custom.py:92]: yes

**Pull request merged**